### ISSUE TITLE
Cannot process: inner whitespace is not related to an open/standalone tag

### ISSUE THREAD

**Issue by oldelvet** [thymeleaf/thymeleaf#467]
Testing Thymeleaf 3 against my existing 2.1 based project I found a number of failures due to failure to parse the template markup. These all work against 2.1 and form a significant barrier to migrating to Thymeleaf 3. I also believe that the actual whitespace involved is valid in XML/HTML and the W3C validators seem to confirm that.

The specific failure seen is due to whitespace in the closing tag of an element after the tag name. i.e.

`</th:block >` instead of `</th:block>`

Any white space (newline etc. causes this and it seems to apply to standard HTML markup as well as thymeleaf specific content).

I checked the template against the W3C XML & HTML validators and they say that the template content is well-formed.

I created a simple example using the thymeleaf 3 sample application and the appropriate commit can be found at
https://github.com/oldelvet/thymeleaf3-template-modes-example/commit/8501338b69434b58620995560b0dd8a79357ab5c

Full stack trace is as follows:

```
Feb 27, 2016 12:30:32 PM org.apache.catalina.core.StandardWrapperValve invoke
SEVERE: Servlet.service() for servlet [dispatcher] in context with path [/thymeleaf3-template-modes-example] threw exception [Request processing failed; nested exception is org.thymeleaf.exceptions.TemplateInputException: An error happened during template parsing (template: "ServletContext resource [/WEB-INF/templates/index.html]")] with root cause
org.thymeleaf.exceptions.TemplateProcessingException: Cannot process: inner whitespace is not related to an open/standalone tag (template: "index.html" - line 17, col 34)
    at org.thymeleaf.engine.TemplateHandlerAdapterMarkupHandler.handleInnerWhiteSpace(TemplateHandlerAdapterMarkupHandler.java:486)
    at org.attoparser.AbstractChainedMarkupHandler.handleInnerWhiteSpace(AbstractChainedMarkupHandler.java:341)
    at org.attoparser.HtmlElement.handleInnerWhiteSpace(HtmlElement.java:268)
    at org.attoparser.HtmlMarkupHandler.handleInnerWhiteSpace(HtmlMarkupHandler.java:550)
    at org.attoparser.AbstractChainedMarkupHandler.handleInnerWhiteSpace(AbstractChainedMarkupHandler.java:341)
    at org.attoparser.ParsingElementMarkupUtil.parseCloseElement(ParsingElementMarkupUtil.java:231)
    at org.attoparser.MarkupParser.parseBuffer(MarkupParser.java:725)
    at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:301)
    at org.attoparser.MarkupParser.parse(MarkupParser.java:257)
    at org.thymeleaf.templateparser.markup.AbstractMarkupTemplateParser.parse(AbstractMarkupTemplateParser.java:233)
    at org.thymeleaf.templateparser.markup.AbstractMarkupTemplateParser.parseStandalone(AbstractMarkupTemplateParser.java:100)
    at org.thymeleaf.engine.TemplateManager.parseAndProcess(TemplateManager.java:662)
    at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1089)
    at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1063)
    at org.thymeleaf.spring4.view.ThymeleafView.renderFragment(ThymeleafView.java:335)
    at org.thymeleaf.spring4.view.ThymeleafView.render(ThymeleafView.java:189)
    at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1244)
    at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1027)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:971)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
    at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:621)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:728)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:305)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:243)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:222)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:123)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:100)
    at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:953)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:118)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1041)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:603)
    at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:310)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by danielfernandez** [comment]
Confirmed. `TemplateHandlerAdapterMarkupHandler` is throwing an exception as soon as an inner whitespace arrives without an open or standalone tag being currently open. It is not attoparser-related, though, it is in the way thymeleaf converts events coming from attoparser to thymeleaf-based events.

**Comment by danielfernandez** [comment]
This fix is already in `3.0.0-SNAPSHOT`
